bin/varnishd/cache/cache_req_fsm.c

```diff
@@ -561,6 +561,21 @@ cnt_lookup(struct worker *wrk, struct req *req)
 		had_objhead = 1;
 	wrk->strangelove = 0;
 	lr = HSH_Lookup(req, &oc, &busy);
+	if (lr == HSH_WALKAWAY) {
+		VRY_Finish(req, DISCARD);
+		AZ(req->objcore);
+		AZ(req->stale_oc);
+		VSLb_ts_req(req, "Waitinglist", W_TIM_real(wrk));
+		VSLb(req->vsl, SLT_Error, "The client is going away");
+		if (req->esi_level > 0) {
```

Comment:
First, we can add a dedicated `HSH_TIMEOUT`. Otherwise I can use the same argument for the error message preventing us to use this branch for future timeouts. I think we will be fine if we dedicated `HSH_WALKAWAY` to the case where the client went away while a request was in a waiting list.

For the timeout case, we  probably want to have different semantics than plainly ending the transaction. Maybe a waiting list timeout would be a deferred `hash_ignore_busy`?

I must admit I don't remember precisely why we did that downstream. I think we needed this to interrupt ESI delivery and prevent it from recursing as the client is gone anyway.

Reply:
There's actually a comment a few lines below explaining why we fail sub-requests:

```c
/* NB: Interrupt delivery upwards to avoid engaging
 * new sub-request tasks.
 */
```

Granted, it doesn't give many details.